{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0f4abada-4017-49ab-b24f-07803f349dda",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!pip install pandas \n",
    "#!pip install numpy \n",
    "#!pip install scikit-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8a2e683-c580-494b-9d0b-4dc5bf69223a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install matplotlib\n",
    "#!pip install seaborn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0791d196-ebdb-4df9-8d23-cdd8f47a545a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install openpyxl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a2cd9756-6af7-4bb1-b784-5d72b1cf07df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "\n",
    "import numpy as np \n",
    "\n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "\n",
    "from sklearn.metrics import mean_squared_error, r2_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e12ec12c-8276-44b6-815a-580e805f145f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC Dataset shape: (242035, 19)\n",
      "Cell Line Details shape: (1002, 13)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ASUS\\OneDrive\\Desktop\\CanScan\\venv\\Lib\\site-packages\\openpyxl\\worksheet\\_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "gdsc_df = pd.read_csv(\"data/GDSC_DATASET.csv\")\n",
    "\n",
    "cell_lines_df = pd.read_excel(\"data/Cell_Lines_Details.xlsx\")\n",
    "\n",
    "print(\"GDSC Dataset shape:\", gdsc_df.shape)\n",
    "\n",
    "print(\"Cell Line Details shape:\", cell_lines_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "061e8182-905f-40f4-a6f5-901b2e751336",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC Dataset:\n",
      "   COSMIC_ID CELL_LINE_NAME     TCGA_DESC  DRUG_ID     DRUG_NAME   LN_IC50  \\\n",
      "0     683667         PFSK-1            MB     1003  Camptothecin -1.463887   \n",
      "1     684057            ES5  UNCLASSIFIED     1003  Camptothecin -3.360586   \n",
      "2     684059            ES7  UNCLASSIFIED     1003  Camptothecin -5.044940   \n",
      "3     684062          EW-11  UNCLASSIFIED     1003  Camptothecin -3.741991   \n",
      "4     684072        SK-ES-1  UNCLASSIFIED     1003  Camptothecin -5.142961   \n",
      "\n",
      "        AUC   Z_SCORE GDSC Tissue descriptor 1 GDSC Tissue descriptor 2  \\\n",
      "0  0.930220  0.433123           nervous_system          medulloblastoma   \n",
      "1  0.791072 -0.599569                     bone           ewings_sarcoma   \n",
      "2  0.592660 -1.516647                     bone           ewings_sarcoma   \n",
      "3  0.734047 -0.807232                     bone           ewings_sarcoma   \n",
      "4  0.582439 -1.570016                     bone           ewings_sarcoma   \n",
      "\n",
      "  Cancer Type (matching TCGA label) Microsatellite instability Status (MSI)  \\\n",
      "0                                MB                               MSS/MSI-L   \n",
      "1                               NaN                               MSS/MSI-L   \n",
      "2                               NaN                               MSS/MSI-L   \n",
      "3                               NaN                               MSS/MSI-L   \n",
      "4                               NaN                               MSS/MSI-L   \n",
      "\n",
      "  Screen Medium Growth Properties CNA Gene Expression Methylation TARGET  \\\n",
      "0             R          Adherent   Y               Y           Y   TOP1   \n",
      "1             R          Adherent   Y               Y           Y   TOP1   \n",
      "2             R          Adherent   Y               Y           Y   TOP1   \n",
      "3             R          Adherent   Y               Y           Y   TOP1   \n",
      "4             R     Semi-Adherent   Y               Y           Y   TOP1   \n",
      "\n",
      "    TARGET_PATHWAY  \n",
      "0  DNA replication  \n",
      "1  DNA replication  \n",
      "2  DNA replication  \n",
      "3  DNA replication  \n",
      "4  DNA replication  \n",
      "\n",
      " Cell Line Details:\n",
      "  Sample Name  COSMIC identifier Whole Exome Sequencing (WES)  \\\n",
      "0        A253           906794.0                            Y   \n",
      "1    BB30-HNC           753531.0                            Y   \n",
      "2    BB49-HNC           753532.0                            Y   \n",
      "3         BHY           753535.0                            Y   \n",
      "4      BICR10          1290724.0                            Y   \n",
      "\n",
      "  Copy Number Alterations (CNA) Gene Expression Methylation Drug\\nResponse  \\\n",
      "0                             Y               Y           Y              Y   \n",
      "1                             Y               Y           Y              Y   \n",
      "2                             Y               Y           Y              Y   \n",
      "3                             Y               Y           Y              Y   \n",
      "4                             Y               Y           Y              Y   \n",
      "\n",
      "  GDSC\\nTissue descriptor 1 GDSC\\nTissue\\ndescriptor 2  \\\n",
      "0            aero_dig_tract              head and neck   \n",
      "1            aero_dig_tract              head and neck   \n",
      "2            aero_dig_tract              head and neck   \n",
      "3            aero_dig_tract              head and neck   \n",
      "4            aero_dig_tract              head and neck   \n",
      "\n",
      "  Cancer Type\\n(matching TCGA label)  \\\n",
      "0                                NaN   \n",
      "1                               HNSC   \n",
      "2                               HNSC   \n",
      "3                               HNSC   \n",
      "4                               HNSC   \n",
      "\n",
      "  Microsatellite \\ninstability Status (MSI) Screen Medium Growth Properties  \n",
      "0                                 MSS/MSI-L         D/F12          Adherent  \n",
      "1                                 MSS/MSI-L         D/F12          Adherent  \n",
      "2                                 MSS/MSI-L         D/F12          Adherent  \n",
      "3                                 MSS/MSI-L         D/F12          Adherent  \n",
      "4                                 MSS/MSI-L         D/F12          Adherent  \n",
      "\n",
      " GDSC Dataset Columns: ['COSMIC_ID', 'CELL_LINE_NAME', 'TCGA_DESC', 'DRUG_ID', 'DRUG_NAME', 'LN_IC50', 'AUC', 'Z_SCORE', 'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2', 'Cancer Type (matching TCGA label)', 'Microsatellite instability Status (MSI)', 'Screen Medium', 'Growth Properties', 'CNA', 'Gene Expression', 'Methylation', 'TARGET', 'TARGET_PATHWAY']\n",
      "\n",
      " Cell Line Metadata Columns: ['Sample Name', 'COSMIC identifier', 'Whole Exome Sequencing (WES)', 'Copy Number Alterations (CNA)', 'Gene Expression', 'Methylation', 'Drug\\nResponse', 'GDSC\\nTissue descriptor 1', 'GDSC\\nTissue\\ndescriptor 2', 'Cancer Type\\n(matching TCGA label)', 'Microsatellite \\ninstability Status (MSI)', 'Screen Medium', 'Growth Properties']\n"
     ]
    }
   ],
   "source": [
    "print(\"GDSC Dataset:\")\n",
    "\n",
    "print(gdsc_df.head())\n",
    "\n",
    "print(\"\\n Cell Line Details:\")\n",
    "\n",
    "print(cell_lines_df.head())\n",
    "\n",
    "print(\"\\n GDSC Dataset Columns:\", gdsc_df.columns.tolist())\n",
    "\n",
    "print(\"\\n Cell Line Metadata Columns:\", cell_lines_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2bcc11b8-e6ce-4bb1-b820-4ac7dd97a7db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC Dataset Preview:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>GDSC Tissue descriptor 1</th>\n",
       "      <th>GDSC Tissue descriptor 2</th>\n",
       "      <th>Cancer Type (matching TCGA label)</th>\n",
       "      <th>Microsatellite instability Status (MSI)</th>\n",
       "      <th>Screen Medium</th>\n",
       "      <th>Growth Properties</th>\n",
       "      <th>CNA</th>\n",
       "      <th>Gene Expression</th>\n",
       "      <th>Methylation</th>\n",
       "      <th>TARGET</th>\n",
       "      <th>TARGET_PATHWAY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>683667</td>\n",
       "      <td>PFSK-1</td>\n",
       "      <td>MB</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-1.463887</td>\n",
       "      <td>0.930220</td>\n",
       "      <td>0.433123</td>\n",
       "      <td>nervous_system</td>\n",
       "      <td>medulloblastoma</td>\n",
       "      <td>MB</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-3.360586</td>\n",
       "      <td>0.791072</td>\n",
       "      <td>-0.599569</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>684059</td>\n",
       "      <td>ES7</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-5.044940</td>\n",
       "      <td>0.592660</td>\n",
       "      <td>-1.516647</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>684062</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-3.741991</td>\n",
       "      <td>0.734047</td>\n",
       "      <td>-0.807232</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>684072</td>\n",
       "      <td>SK-ES-1</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-5.142961</td>\n",
       "      <td>0.582439</td>\n",
       "      <td>-1.570016</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Semi-Adherent</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   COSMIC_ID CELL_LINE_NAME     TCGA_DESC  DRUG_ID     DRUG_NAME   LN_IC50  \\\n",
       "0     683667         PFSK-1            MB     1003  Camptothecin -1.463887   \n",
       "1     684057            ES5  UNCLASSIFIED     1003  Camptothecin -3.360586   \n",
       "2     684059            ES7  UNCLASSIFIED     1003  Camptothecin -5.044940   \n",
       "3     684062          EW-11  UNCLASSIFIED     1003  Camptothecin -3.741991   \n",
       "4     684072        SK-ES-1  UNCLASSIFIED     1003  Camptothecin -5.142961   \n",
       "\n",
       "        AUC   Z_SCORE GDSC Tissue descriptor 1 GDSC Tissue descriptor 2  \\\n",
       "0  0.930220  0.433123           nervous_system          medulloblastoma   \n",
       "1  0.791072 -0.599569                     bone           ewings_sarcoma   \n",
       "2  0.592660 -1.516647                     bone           ewings_sarcoma   \n",
       "3  0.734047 -0.807232                     bone           ewings_sarcoma   \n",
       "4  0.582439 -1.570016                     bone           ewings_sarcoma   \n",
       "\n",
       "  Cancer Type (matching TCGA label) Microsatellite instability Status (MSI)  \\\n",
       "0                                MB                               MSS/MSI-L   \n",
       "1                               NaN                               MSS/MSI-L   \n",
       "2                               NaN                               MSS/MSI-L   \n",
       "3                               NaN                               MSS/MSI-L   \n",
       "4                               NaN                               MSS/MSI-L   \n",
       "\n",
       "  Screen Medium Growth Properties CNA Gene Expression Methylation TARGET  \\\n",
       "0             R          Adherent   Y               Y           Y   TOP1   \n",
       "1             R          Adherent   Y               Y           Y   TOP1   \n",
       "2             R          Adherent   Y               Y           Y   TOP1   \n",
       "3             R          Adherent   Y               Y           Y   TOP1   \n",
       "4             R     Semi-Adherent   Y               Y           Y   TOP1   \n",
       "\n",
       "    TARGET_PATHWAY  \n",
       "0  DNA replication  \n",
       "1  DNA replication  \n",
       "2  DNA replication  \n",
       "3  DNA replication  \n",
       "4  DNA replication  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Cell Line Details Preview:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th>Whole Exome Sequencing (WES)</th>\n",
       "      <th>Copy Number Alterations (CNA)</th>\n",
       "      <th>Gene Expression</th>\n",
       "      <th>Methylation</th>\n",
       "      <th>Drug\\nResponse</th>\n",
       "      <th>GDSC\\nTissue descriptor 1</th>\n",
       "      <th>GDSC\\nTissue\\ndescriptor 2</th>\n",
       "      <th>Cancer Type\\n(matching TCGA label)</th>\n",
       "      <th>Microsatellite \\ninstability Status (MSI)</th>\n",
       "      <th>Screen Medium</th>\n",
       "      <th>Growth Properties</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A253</td>\n",
       "      <td>906794.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BB30-HNC</td>\n",
       "      <td>753531.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BB49-HNC</td>\n",
       "      <td>753532.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BHY</td>\n",
       "      <td>753535.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BICR10</td>\n",
       "      <td>1290724.0</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td>HNSC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample Name  COSMIC identifier Whole Exome Sequencing (WES)  \\\n",
       "0        A253           906794.0                            Y   \n",
       "1    BB30-HNC           753531.0                            Y   \n",
       "2    BB49-HNC           753532.0                            Y   \n",
       "3         BHY           753535.0                            Y   \n",
       "4      BICR10          1290724.0                            Y   \n",
       "\n",
       "  Copy Number Alterations (CNA) Gene Expression Methylation Drug\\nResponse  \\\n",
       "0                             Y               Y           Y              Y   \n",
       "1                             Y               Y           Y              Y   \n",
       "2                             Y               Y           Y              Y   \n",
       "3                             Y               Y           Y              Y   \n",
       "4                             Y               Y           Y              Y   \n",
       "\n",
       "  GDSC\\nTissue descriptor 1 GDSC\\nTissue\\ndescriptor 2  \\\n",
       "0            aero_dig_tract              head and neck   \n",
       "1            aero_dig_tract              head and neck   \n",
       "2            aero_dig_tract              head and neck   \n",
       "3            aero_dig_tract              head and neck   \n",
       "4            aero_dig_tract              head and neck   \n",
       "\n",
       "  Cancer Type\\n(matching TCGA label)  \\\n",
       "0                                NaN   \n",
       "1                               HNSC   \n",
       "2                               HNSC   \n",
       "3                               HNSC   \n",
       "4                               HNSC   \n",
       "\n",
       "  Microsatellite \\ninstability Status (MSI) Screen Medium Growth Properties  \n",
       "0                                 MSS/MSI-L         D/F12          Adherent  \n",
       "1                                 MSS/MSI-L         D/F12          Adherent  \n",
       "2                                 MSS/MSI-L         D/F12          Adherent  \n",
       "3                                 MSS/MSI-L         D/F12          Adherent  \n",
       "4                                 MSS/MSI-L         D/F12          Adherent  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(\"GDSC Dataset Preview:\")\n",
    "\n",
    "display(gdsc_df.head())\n",
    "\n",
    "print(\"\\n Cell Line Details Preview:\")\n",
    "\n",
    "display(cell_lines_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "46256b7d-48ba-4f5c-87dc-d81ef2801ca5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC Dataset Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 242035 entries, 0 to 242034\n",
      "Data columns (total 19 columns):\n",
      " #   Column                                   Non-Null Count   Dtype  \n",
      "---  ------                                   --------------   -----  \n",
      " 0   COSMIC_ID                                242035 non-null  int64  \n",
      " 1   CELL_LINE_NAME                           242035 non-null  object \n",
      " 2   TCGA_DESC                                240968 non-null  object \n",
      " 3   DRUG_ID                                  242035 non-null  int64  \n",
      " 4   DRUG_NAME                                242035 non-null  object \n",
      " 5   LN_IC50                                  242035 non-null  float64\n",
      " 6   AUC                                      242035 non-null  float64\n",
      " 7   Z_SCORE                                  242035 non-null  float64\n",
      " 8   GDSC Tissue descriptor 1                 232669 non-null  object \n",
      " 9   GDSC Tissue descriptor 2                 232669 non-null  object \n",
      " 10  Cancer Type (matching TCGA label)        190589 non-null  object \n",
      " 11  Microsatellite instability Status (MSI)  229682 non-null  object \n",
      " 12  Screen Medium                            232669 non-null  object \n",
      " 13  Growth Properties                        232669 non-null  object \n",
      " 14  CNA                                      232669 non-null  object \n",
      " 15  Gene Expression                          232669 non-null  object \n",
      " 16  Methylation                              232669 non-null  object \n",
      " 17  TARGET                                   214880 non-null  object \n",
      " 18  TARGET_PATHWAY                           242035 non-null  object \n",
      "dtypes: float64(3), int64(2), object(14)\n",
      "memory usage: 35.1+ MB\n",
      "\n",
      " Missing values in GDSC:\n",
      "COSMIC_ID                                      0\n",
      "CELL_LINE_NAME                                 0\n",
      "TCGA_DESC                                   1067\n",
      "DRUG_ID                                        0\n",
      "DRUG_NAME                                      0\n",
      "LN_IC50                                        0\n",
      "AUC                                            0\n",
      "Z_SCORE                                        0\n",
      "GDSC Tissue descriptor 1                    9366\n",
      "GDSC Tissue descriptor 2                    9366\n",
      "Cancer Type (matching TCGA label)          51446\n",
      "Microsatellite instability Status (MSI)    12353\n",
      "Screen Medium                               9366\n",
      "Growth Properties                           9366\n",
      "CNA                                         9366\n",
      "Gene Expression                             9366\n",
      "Methylation                                 9366\n",
      "TARGET                                     27155\n",
      "TARGET_PATHWAY                                 0\n",
      "dtype: int64\n",
      "\n",
      " Cell Line Details Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1002 entries, 0 to 1001\n",
      "Data columns (total 13 columns):\n",
      " #   Column                                    Non-Null Count  Dtype  \n",
      "---  ------                                    --------------  -----  \n",
      " 0   Sample Name                               1002 non-null   object \n",
      " 1   COSMIC identifier                         1001 non-null   float64\n",
      " 2   Whole Exome Sequencing (WES)              1002 non-null   object \n",
      " 3   Copy Number Alterations (CNA)             1002 non-null   object \n",
      " 4   Gene Expression                           1002 non-null   object \n",
      " 5   Methylation                               1002 non-null   object \n",
      " 6   Drug\n",
      "Response                             1002 non-null   object \n",
      " 7   GDSC\n",
      "Tissue descriptor 1                  1001 non-null   object \n",
      " 8   GDSC\n",
      "Tissue\n",
      "descriptor 2                  1001 non-null   object \n",
      " 9   Cancer Type\n",
      "(matching TCGA label)         826 non-null    object \n",
      " 10  Microsatellite \n",
      "instability Status (MSI)  986 non-null    object \n",
      " 11  Screen Medium                             1001 non-null   object \n",
      " 12  Growth Properties                         999 non-null    object \n",
      "dtypes: float64(1), object(12)\n",
      "memory usage: 101.9+ KB\n",
      "\n",
      " Missing values in Cell Line Details:\n",
      "Sample Name                                    0\n",
      "COSMIC identifier                              1\n",
      "Whole Exome Sequencing (WES)                   0\n",
      "Copy Number Alterations (CNA)                  0\n",
      "Gene Expression                                0\n",
      "Methylation                                    0\n",
      "Drug\\nResponse                                 0\n",
      "GDSC\\nTissue descriptor 1                      1\n",
      "GDSC\\nTissue\\ndescriptor 2                     1\n",
      "Cancer Type\\n(matching TCGA label)           176\n",
      "Microsatellite \\ninstability Status (MSI)     16\n",
      "Screen Medium                                  1\n",
      "Growth Properties                              3\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"GDSC Dataset Info:\")\n",
    "\n",
    "gdsc_df.info()\n",
    "\n",
    "print(\"\\n Missing values in GDSC:\")\n",
    "\n",
    "print(gdsc_df.isnull().sum())\n",
    "\n",
    "print(\"\\n Cell Line Details Info:\")\n",
    "\n",
    "cell_lines_df.info()\n",
    "\n",
    "print(\"\\n Missing values in Cell Line Details:\")\n",
    "\n",
    "print(cell_lines_df.isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "204ec3c1-cea6-4aa0-aadd-98963e22cef2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged dataset shape: (242035, 31)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>GDSC Tissue descriptor 1</th>\n",
       "      <th>GDSC Tissue descriptor 2</th>\n",
       "      <th>...</th>\n",
       "      <th>Copy Number Alterations (CNA)</th>\n",
       "      <th>Gene Expression_y</th>\n",
       "      <th>Methylation_y</th>\n",
       "      <th>DrugResponse</th>\n",
       "      <th>GDSCTissue descriptor 1</th>\n",
       "      <th>GDSCTissuedescriptor 2</th>\n",
       "      <th>Cancer Type(matching TCGA label)</th>\n",
       "      <th>Microsatellite instability Status (MSI)_y</th>\n",
       "      <th>Screen Medium_y</th>\n",
       "      <th>Growth Properties_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>683667</td>\n",
       "      <td>PFSK-1</td>\n",
       "      <td>MB</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-1.463887</td>\n",
       "      <td>0.930220</td>\n",
       "      <td>0.433123</td>\n",
       "      <td>nervous_system</td>\n",
       "      <td>medulloblastoma</td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>nervous_system</td>\n",
       "      <td>medulloblastoma</td>\n",
       "      <td>MB</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-3.360586</td>\n",
       "      <td>0.791072</td>\n",
       "      <td>-0.599569</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>684059</td>\n",
       "      <td>ES7</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-5.044940</td>\n",
       "      <td>0.592660</td>\n",
       "      <td>-1.516647</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>684062</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-3.741991</td>\n",
       "      <td>0.734047</td>\n",
       "      <td>-0.807232</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>684072</td>\n",
       "      <td>SK-ES-1</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>-5.142961</td>\n",
       "      <td>0.582439</td>\n",
       "      <td>-1.570016</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>bone</td>\n",
       "      <td>ewings_sarcoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>Semi-Adherent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   COSMIC_ID CELL_LINE_NAME     TCGA_DESC  DRUG_ID     DRUG_NAME   LN_IC50  \\\n",
       "0     683667         PFSK-1            MB     1003  Camptothecin -1.463887   \n",
       "1     684057            ES5  UNCLASSIFIED     1003  Camptothecin -3.360586   \n",
       "2     684059            ES7  UNCLASSIFIED     1003  Camptothecin -5.044940   \n",
       "3     684062          EW-11  UNCLASSIFIED     1003  Camptothecin -3.741991   \n",
       "4     684072        SK-ES-1  UNCLASSIFIED     1003  Camptothecin -5.142961   \n",
       "\n",
       "        AUC   Z_SCORE GDSC Tissue descriptor 1 GDSC Tissue descriptor 2  ...  \\\n",
       "0  0.930220  0.433123           nervous_system          medulloblastoma  ...   \n",
       "1  0.791072 -0.599569                     bone           ewings_sarcoma  ...   \n",
       "2  0.592660 -1.516647                     bone           ewings_sarcoma  ...   \n",
       "3  0.734047 -0.807232                     bone           ewings_sarcoma  ...   \n",
       "4  0.582439 -1.570016                     bone           ewings_sarcoma  ...   \n",
       "\n",
       "  Copy Number Alterations (CNA) Gene Expression_y Methylation_y DrugResponse  \\\n",
       "0                             Y                 Y             Y            Y   \n",
       "1                             Y                 Y             Y            Y   \n",
       "2                             Y                 Y             Y            Y   \n",
       "3                             Y                 Y             Y            Y   \n",
       "4                             Y                 Y             Y            Y   \n",
       "\n",
       "  GDSCTissue descriptor 1 GDSCTissuedescriptor 2  \\\n",
       "0          nervous_system        medulloblastoma   \n",
       "1                    bone         ewings_sarcoma   \n",
       "2                    bone         ewings_sarcoma   \n",
       "3                    bone         ewings_sarcoma   \n",
       "4                    bone         ewings_sarcoma   \n",
       "\n",
       "  Cancer Type(matching TCGA label) Microsatellite instability Status (MSI)_y  \\\n",
       "0                               MB                                 MSS/MSI-L   \n",
       "1                              NaN                                 MSS/MSI-L   \n",
       "2                              NaN                                 MSS/MSI-L   \n",
       "3                              NaN                                 MSS/MSI-L   \n",
       "4                              NaN                                 MSS/MSI-L   \n",
       "\n",
       "  Screen Medium_y Growth Properties_y  \n",
       "0               R            Adherent  \n",
       "1               R            Adherent  \n",
       "2               R            Adherent  \n",
       "3               R            Adherent  \n",
       "4               R       Semi-Adherent  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_lines_df.columns=[col.replace('\\n', '').replace('','').strip() for col in cell_lines_df.columns]\n",
    "\n",
    "cell_lines_df.rename(columns={\n",
    "\n",
    "    'COSMIC identifier': 'COSMIC_ID',\n",
    "\n",
    "    'Cancer TYPE (matching TCGA label)': 'Cancer Type (matching TCGA label)',\n",
    "\n",
    "}, inplace=True)\n",
    "\n",
    "cell_lines_df.dropna(subset=['COSMIC_ID'], inplace=True)\n",
    "\n",
    "merged_df=pd.merge(gdsc_df, cell_lines_df, on='COSMIC_ID', how='left')\n",
    "\n",
    "print(\"Merged dataset shape:\", merged_df.shape)\n",
    "\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6dcfbb92-5612-4856-be8d-78cbda1052ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Preprocessing complete. Final dataset shape: (228615, 18)\n"
     ]
    }
   ],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "columns_to_drop = [\n",
    "    'COSMIC_ID', 'CELL_LINE_NAME', 'DRUG_ID', 'DRUG_NAME',\n",
    "    \n",
    "    'Gene Expression_y', 'Copy Number Alterations (CNA)', 'Methylation_y',\n",
    "    \n",
    "    'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2',\n",
    "    \n",
    "    'Cancer Type (matching TCGA label)', \n",
    "    \n",
    "    'Microsatellite instability Status (MSI)',  \n",
    "    \n",
    "    'Screen Medium', 'Growth Properties'\n",
    "    \n",
    "]\n",
    "\n",
    "merged_df.drop(columns=columns_to_drop, inplace=True, errors='ignore')\n",
    "\n",
    "merged_df.rename(columns={\n",
    "    'Cancer Type(matching TCGA label)': 'Cancer_Type',\n",
    "    \n",
    "    'Microsatellite instability Status (MSI)_y': 'MSI_Status',\n",
    "    \n",
    "    'Screen Medium_y': 'Screen_Medium',\n",
    "    \n",
    "    'Growth Properties_y': 'Growth_Properties'\n",
    "    \n",
    "}, inplace=True)\n",
    "\n",
    "for col in ['Gene Expression_x', 'Methylation_x', 'DrugResponse']:\n",
    "    \n",
    "    merged_df[col] = merged_df[col].map({'Y': 1, 'N': 0})\n",
    "\n",
    "label_encoders = {}\n",
    "\n",
    "for col in ['Cancer_Type', 'MSI_Status', 'Screen_Medium', 'Growth_Properties', 'TARGET', 'TARGET_PATHWAY']:\n",
    "    \n",
    "    le = LabelEncoder()\n",
    "    \n",
    "    merged_df[col] = merged_df[col].fillna(\"Unknown\")\n",
    "    \n",
    "    merged_df[col] = le.fit_transform(merged_df[col])\n",
    "    \n",
    "    label_encoders[col] = le  \n",
    "\n",
    "merged_df.dropna(inplace=True)\n",
    "\n",
    "X = merged_df.drop(columns=['LN_IC50', 'AUC', 'Z_SCORE']) \n",
    "\n",
    "y = merged_df['LN_IC50']\n",
    "\n",
    "print(\" Preprocessing complete. Final dataset shape:\", X.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41960d5d-f189-4df4-9301-e0aac7d3380a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "801f9f4c-4550-44b1-89ec-d770bbee8824",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non-numeric columns in X: ['TCGA_DESC', 'Microsatellite instability Status (MSI)_x', 'Screen Medium_x', 'Growth Properties_x', 'CNA', 'Sample Name', 'Whole Exome Sequencing (WES)', 'GDSCTissue descriptor 1', 'GDSCTissuedescriptor 2']\n"
     ]
    }
   ],
   "source": [
    "non_numeric_cols = X.select_dtypes(include=['object']).columns.tolist()\n",
    "print(\"Non-numeric columns in X:\", non_numeric_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fd66e22a-f078-4df0-96f5-dd4a0d2de8bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "cols_to_drop = [\n",
    "    'TCGA_DESC', \n",
    "    'Screen Medium_x', \n",
    "    'Growth Properties_x',\n",
    "    'Sample Name'\n",
    "]\n",
    "\n",
    "X.drop(columns=cols_to_drop, inplace=True, errors='ignore')\n",
    "\n",
    "if 'Whole Exome Sequencing (WES)' in X.columns:\n",
    "    \n",
    "    X['Whole Exome Sequencing (WES)'] = X['Whole Exome Sequencing (WES)'].map({'Y': 1, 'N': 0, 'MB': 2}).fillna(0)\n",
    "\n",
    "label_cols = [\n",
    "    'Microsatellite instability Status (MSI)_x',\n",
    "    'CNA',\n",
    "    'GDSCTissue descriptor 1',\n",
    "    'GDSCTissuedescriptor 2'\n",
    "]\n",
    "\n",
    "le_dict = {}\n",
    "\n",
    "for col in label_cols:\n",
    "    \n",
    "    le = LabelEncoder()\n",
    "    \n",
    "    X[col] = le.fit_transform(X[col].astype(str).fillna(\"Unknown\"))\n",
    "    \n",
    "    le_dict[col] = le"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fa474cb1-7e3f-4fdf-b5ae-590a076d0edb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "# Ensure all data is float32\n",
    "X = X.astype('float32')\n",
    "\n",
    "y = y.astype('float32')\n",
    "\n",
    "# Step 1: Split data\n",
    "X_train, X_temp, y_train, y_temp = train_test_split(\n",
    "    \n",
    "    X, y, test_size=0.3, random_state=42\n",
    ")\n",
    "\n",
    "X_val, X_test, y_val, y_test = train_test_split(\n",
    "    \n",
    "    X_temp, y_temp, test_size=0.5, random_state=42\n",
    ")\n",
    "\n",
    "# Step 2: Scale input features\n",
    "X_scaler = StandardScaler()\n",
    "\n",
    "X_train_scaled = X_scaler.fit_transform(X_train)\n",
    "\n",
    "X_val_scaled   = X_scaler.transform(X_val)\n",
    "\n",
    "X_test_scaled  = X_scaler.transform(X_test)\n",
    "\n",
    "# Step 3: Scale target variable (reshape needed!)\n",
    "y_scaler = StandardScaler()\n",
    "\n",
    "y_train_scaled = y_scaler.fit_transform(y_train.values.reshape(-1, 1))\n",
    "\n",
    "y_val_scaled   = y_scaler.transform(y_val.values.reshape(-1, 1))\n",
    "\n",
    "y_test_scaled  = y_scaler.transform(y_test.values.reshape(-1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cb740ea4-8e03-455d-ba65-32895d0dd02c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "\n",
    "# Convert to tensors\n",
    "X_train_tensor = torch.tensor(X_train_scaled, dtype=torch.float32)\n",
    "\n",
    "X_val_tensor = torch.tensor(X_val_scaled, dtype=torch.float32)\n",
    "\n",
    "X_test_tensor = torch.tensor(X_test_scaled, dtype=torch.float32)\n",
    "\n",
    "y_train_tensor = torch.tensor(y_train_scaled, dtype=torch.float32)\n",
    "\n",
    "y_val_tensor = torch.tensor(y_val_scaled, dtype=torch.float32)\n",
    "\n",
    "y_test_tensor = torch.tensor(y_test_scaled, dtype=torch.float32)\n",
    "\n",
    "# Create DataLoader\n",
    "batch_size = 256\n",
    "\n",
    "train_dataset = TensorDataset(X_train_tensor, y_train_tensor)\n",
    "\n",
    "val_dataset = TensorDataset(X_val_tensor, y_val_tensor)\n",
    "\n",
    "train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)\n",
    "\n",
    "val_loader = DataLoader(val_dataset, batch_size=batch_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dc803af2-1861-4e39-bff0-22275cbe86d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch.nn as nn\n",
    "\n",
    "class DeepCancerDrugPredictor(nn.Module):\n",
    "    \n",
    "    def __init__(self, input_dim):\n",
    "        \n",
    "        super().__init__()\n",
    "        \n",
    "        self.model = nn.Sequential(\n",
    "            nn.Linear(input_dim, 256),\n",
    "            nn.BatchNorm1d(256),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(0.3),\n",
    "            \n",
    "            nn.Linear(256, 128),\n",
    "            nn.BatchNorm1d(128),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(0.3),\n",
    "\n",
    "            nn.Linear(128, 64),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(0.2),\n",
    "\n",
    "            nn.Linear(64, 32),\n",
    "            nn.ReLU(),\n",
    "\n",
    "            nn.Linear(32, 1)\n",
    "        )\n",
    "        \n",
    "    def forward(self, x):\n",
    "        \n",
    "        return self.model(x)\n",
    "        \n",
    "model = DeepCancerDrugPredictor(input_dim=X_train_tensor.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "65707f62-d33e-4022-a954-6049b3965f53",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import torch.optim as optim\n",
    "\n",
    "criterion = nn.MSELoss()\n",
    "\n",
    "optimizer = optim.Adam(model.parameters(), lr=0.001)\n",
    "\n",
    "scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', patience=3, factor=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a90a6b30-7e26-482a-842c-6760f8c7dcef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1, Train Loss: 0.9522, Val Loss: 0.9017\n",
      "Epoch 2, Train Loss: 0.9092, Val Loss: 0.8556\n",
      "Epoch 3, Train Loss: 0.8764, Val Loss: 0.7880\n",
      "Epoch 4, Train Loss: 0.8464, Val Loss: 0.7562\n",
      "Epoch 5, Train Loss: 0.8238, Val Loss: 0.7768\n",
      "Epoch 6, Train Loss: 0.8107, Val Loss: 0.7242\n",
      "Epoch 7, Train Loss: 0.7935, Val Loss: 0.7022\n",
      "Epoch 8, Train Loss: 0.7863, Val Loss: 0.6776\n",
      "Epoch 9, Train Loss: 0.7768, Val Loss: 0.6703\n",
      "Epoch 10, Train Loss: 0.7656, Val Loss: 0.6540\n",
      "Epoch 11, Train Loss: 0.7502, Val Loss: 0.6373\n",
      "Epoch 12, Train Loss: 0.7403, Val Loss: 0.6288\n",
      "Epoch 13, Train Loss: 0.7349, Val Loss: 0.6374\n",
      "Epoch 14, Train Loss: 0.7267, Val Loss: 0.6187\n",
      "Epoch 15, Train Loss: 0.7212, Val Loss: 0.6078\n",
      "Epoch 16, Train Loss: 0.7072, Val Loss: 0.6061\n",
      "Epoch 17, Train Loss: 0.7014, Val Loss: 0.6019\n",
      "Epoch 18, Train Loss: 0.6992, Val Loss: 0.5917\n",
      "Epoch 19, Train Loss: 0.6924, Val Loss: 0.5859\n",
      "Epoch 20, Train Loss: 0.6892, Val Loss: 0.5835\n",
      "Epoch 21, Train Loss: 0.6893, Val Loss: 0.5885\n",
      "Epoch 22, Train Loss: 0.6874, Val Loss: 0.5851\n",
      "Epoch 23, Train Loss: 0.6822, Val Loss: 0.5773\n",
      "Epoch 24, Train Loss: 0.6741, Val Loss: 0.5817\n",
      "Epoch 25, Train Loss: 0.6737, Val Loss: 0.5773\n",
      "Epoch 26, Train Loss: 0.6714, Val Loss: 0.5770\n",
      "Epoch 27, Train Loss: 0.6654, Val Loss: 0.5936\n",
      "Epoch 28, Train Loss: 0.6664, Val Loss: 0.5704\n",
      "Epoch 29, Train Loss: 0.6618, Val Loss: 0.5666\n",
      "Epoch 30, Train Loss: 0.6551, Val Loss: 0.5702\n",
      "Epoch 31, Train Loss: 0.6548, Val Loss: 0.5656\n",
      "Epoch 32, Train Loss: 0.6527, Val Loss: 0.5605\n",
      "Epoch 33, Train Loss: 0.6525, Val Loss: 0.5571\n",
      "Epoch 34, Train Loss: 0.6464, Val Loss: 0.5631\n",
      "Epoch 35, Train Loss: 0.6451, Val Loss: 0.5513\n",
      "Epoch 36, Train Loss: 0.6483, Val Loss: 0.5566\n",
      "Epoch 37, Train Loss: 0.6411, Val Loss: 0.5533\n",
      "Epoch 38, Train Loss: 0.6396, Val Loss: 0.5677\n",
      "Epoch 39, Train Loss: 0.6377, Val Loss: 0.5478\n",
      "Epoch 40, Train Loss: 0.6359, Val Loss: 0.5501\n",
      "Epoch 41, Train Loss: 0.6305, Val Loss: 0.5479\n",
      "Epoch 42, Train Loss: 0.6273, Val Loss: 0.5428\n",
      "Epoch 43, Train Loss: 0.6237, Val Loss: 0.5376\n",
      "Epoch 44, Train Loss: 0.6287, Val Loss: 0.5509\n",
      "Epoch 45, Train Loss: 0.6279, Val Loss: 0.5401\n",
      "Epoch 46, Train Loss: 0.6218, Val Loss: 0.5350\n",
      "Epoch 47, Train Loss: 0.6204, Val Loss: 0.5475\n",
      "Epoch 48, Train Loss: 0.6209, Val Loss: 0.5369\n",
      "Epoch 49, Train Loss: 0.6171, Val Loss: 0.5329\n",
      "Epoch 50, Train Loss: 0.6175, Val Loss: 0.5299\n",
      "Epoch 51, Train Loss: 0.6186, Val Loss: 0.5355\n",
      "Epoch 52, Train Loss: 0.6106, Val Loss: 0.5395\n",
      "Epoch 53, Train Loss: 0.6190, Val Loss: 0.5312\n",
      "Epoch 54, Train Loss: 0.6106, Val Loss: 0.5306\n",
      "Epoch 55, Train Loss: 0.5960, Val Loss: 0.5203\n",
      "Epoch 56, Train Loss: 0.6020, Val Loss: 0.5295\n",
      "Epoch 57, Train Loss: 0.5937, Val Loss: 0.5201\n",
      "Epoch 58, Train Loss: 0.5980, Val Loss: 0.5133\n",
      "Epoch 59, Train Loss: 0.5913, Val Loss: 0.5243\n",
      "Epoch 60, Train Loss: 0.5909, Val Loss: 0.5137\n",
      "Epoch 61, Train Loss: 0.5889, Val Loss: 0.5247\n",
      "Epoch 62, Train Loss: 0.5930, Val Loss: 0.5210\n",
      "Epoch 63, Train Loss: 0.5888, Val Loss: 0.5110\n",
      "Epoch 64, Train Loss: 0.5837, Val Loss: 0.5088\n",
      "Epoch 65, Train Loss: 0.5833, Val Loss: 0.5098\n",
      "Epoch 66, Train Loss: 0.5837, Val Loss: 0.5056\n",
      "Epoch 67, Train Loss: 0.5789, Val Loss: 0.5028\n",
      "Epoch 68, Train Loss: 0.5817, Val Loss: 0.5111\n",
      "Epoch 69, Train Loss: 0.5796, Val Loss: 0.5232\n",
      "Epoch 70, Train Loss: 0.5822, Val Loss: 0.5033\n",
      "Epoch 71, Train Loss: 0.5797, Val Loss: 0.5060\n",
      "Epoch 72, Train Loss: 0.5752, Val Loss: 0.4996\n",
      "Epoch 73, Train Loss: 0.5743, Val Loss: 0.5003\n",
      "Epoch 74, Train Loss: 0.5752, Val Loss: 0.5020\n",
      "Epoch 75, Train Loss: 0.5741, Val Loss: 0.5039\n",
      "Epoch 76, Train Loss: 0.5764, Val Loss: 0.5016\n",
      "Epoch 77, Train Loss: 0.5715, Val Loss: 0.5376\n",
      "Early stopping triggered.\n"
     ]
    }
   ],
   "source": [
    "epochs = 100\n",
    "\n",
    "\n",
    "patience = 5\n",
    "\n",
    "best_val_loss = float('inf')\n",
    "\n",
    "counter = 0\n",
    "\n",
    "for epoch in range(epochs):\n",
    "    \n",
    "    model.train()\n",
    "\n",
    "    train_loss = 0.0\n",
    "\n",
    "    for batch_X, batch_y in train_loader:\n",
    "        \n",
    "        optimizer.zero_grad()\n",
    "        \n",
    "        output = model(batch_X)\n",
    "        \n",
    "        loss = criterion(output, batch_y)\n",
    "        \n",
    "        loss.backward()\n",
    "        \n",
    "        optimizer.step()\n",
    "        \n",
    "        train_loss += loss.item() * batch_X.size(0)  # sum over batch\n",
    "\n",
    "    avg_train_loss = train_loss / len(train_loader.dataset)\n",
    "\n",
    "    # Validation\n",
    "    model.eval()\n",
    "    \n",
    "    val_loss = 0.0\n",
    "    \n",
    "    with torch.no_grad():\n",
    "        \n",
    "        for batch_X, batch_y in val_loader:\n",
    "            \n",
    "            val_output = model(batch_X)\n",
    "            \n",
    "            loss = criterion(val_output, batch_y)\n",
    "            \n",
    "            val_loss += loss.item() * batch_X.size(0)\n",
    "\n",
    "    avg_val_loss = val_loss / len(val_loader.dataset)\n",
    "    \n",
    "    print(f\"Epoch {epoch+1}, Train Loss: {avg_train_loss:.4f}, Val Loss: {avg_val_loss:.4f}\")\n",
    "\n",
    "    scheduler.step(avg_val_loss)\n",
    "\n",
    "    if avg_val_loss < best_val_loss:\n",
    "        \n",
    "        best_val_loss = avg_val_loss\n",
    "        \n",
    "        counter = 0\n",
    "        \n",
    "        torch.save(model.state_dict(), 'best_model.pt')\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        counter += 1\n",
    "        \n",
    "        if counter >= patience:\n",
    "            \n",
    "            print(\"Early stopping triggered.\")\n",
    "            \n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ce6e79b3-f80d-4216-b3c1-1f97f578033b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Loss: 0.5029\n"
     ]
    }
   ],
   "source": [
    "model.load_state_dict(torch.load('best_model.pt'))\n",
    "\n",
    "model.eval()\n",
    "\n",
    "with torch.no_grad():\n",
    "    \n",
    "    test_preds = model(X_test_tensor)\n",
    "    \n",
    "    test_loss = criterion(test_preds, y_test_tensor)\n",
    "    \n",
    "print(f\"Test Loss: {test_loss.item():.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7dbbcee7-b969-40e8-8624-2e09c0cca3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_drug_sensitivity(sample_dict, model, scaler_y, le_dict, X_columns):\n",
    "    \n",
    "    import numpy as np\n",
    "    \n",
    "    import pandas as pd\n",
    "    \n",
    "    import torch\n",
    "\n",
    "    model.eval()\n",
    "\n",
    "    # Step 1: Convert the sample dictionary to a DataFrame\n",
    "    df_sample = pd.DataFrame([sample_dict])\n",
    "\n",
    "    # Step 2: Label encode categorical features\n",
    "    for col, le in le_dict.items():\n",
    "        \n",
    "        if col in df_sample.columns:\n",
    "            \n",
    "            val = df_sample[col].astype(str).values[0]\n",
    "            \n",
    "            if val in le.classes_:\n",
    "                \n",
    "                df_sample[col] = le.transform([val])\n",
    "                \n",
    "            elif \"Unknown\" in le.classes_:\n",
    "                \n",
    "                df_sample[col] = le.transform([\"Unknown\"])\n",
    "                \n",
    "            else:\n",
    "                \n",
    "                # fallback to first class (safe option)\n",
    "                df_sample[col] = le.transform([le.classes_[0]])\n",
    "\n",
    "    # Step 3: Handle binary/multi-class mapping manually if needed\n",
    "    if 'Whole Exome Sequencing (WES)' in df_sample.columns:\n",
    "        \n",
    "        df_sample['Whole Exome Sequencing (WES)'] = df_sample['Whole Exome Sequencing (WES)'].map(\n",
    "            \n",
    "            {'Y': 1, 'N': 0, 'MB': 2}).fillna(0)\n",
    "\n",
    "    # Step 4: Reindex to ensure correct order and fill missing features with 0\n",
    "    df_sample = df_sample.reindex(columns=X_columns, fill_value=0)\n",
    "\n",
    "    # Step 5: Convert to tensor\n",
    "    input_tensor = torch.tensor(df_sample.values, dtype=torch.float32)\n",
    "\n",
    "    # Step 6: Model prediction\n",
    "    with torch.no_grad():\n",
    "        \n",
    "        pred_scaled = model(input_tensor).item()\n",
    "\n",
    "    # Step 7: Inverse scale the prediction to original range\n",
    "    pred_original = scaler_y.inverse_transform([[pred_scaled]])[0][0]\n",
    "\n",
    "    return pred_original"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "df0ee0b2-a9ff-4a5a-a39e-4de259d40b0a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Predicted Drug Sensitivity: 6.4271\n"
     ]
    }
   ],
   "source": [
    "# Define a sample input with key features\n",
    "sample_input = {\n",
    "    'Microsatellite instability Status (MSI)_x': 'MSI-High',\n",
    "    'CNA': 'High',\n",
    "    'GDSCTissue descriptor 1': 'Lung',\n",
    "    'GDSCTissuedescriptor 2': 'NSCLC',\n",
    "    'Whole Exome Sequencing (WES)': 'Y',\n",
    "    # Other features (numerical) will be filled with 0 in reindex step of function\n",
    "}\n",
    "\n",
    "# Make prediction\n",
    "prediction = predict_drug_sensitivity(\n",
    "    \n",
    "    sample_dict=sample_input,\n",
    "    \n",
    "    model=model,\n",
    "    \n",
    "    scaler_y=y_scaler,      # This is your trained output scaler\n",
    "    \n",
    "    le_dict=le_dict,        # Label encoders for categorical features\n",
    "    \n",
    "    X_columns=X.columns     # Feature column order used in training\n",
    "    \n",
    ")\n",
    "\n",
    "# Display the result\n",
    "print(f\" Predicted Drug Sensitivity: {prediction:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a022f45e-cf7e-4f0a-8fbc-e2b711819ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Target (y) value range:\n",
      "-8.642551 13.820189\n"
     ]
    }
   ],
   "source": [
    "print(\"Target (y) value range:\")\n",
    "\n",
    "print(y.min(), y.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75db00bb-96e9-4d90-859b-ab73d9e57b4f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
